adc.c

```diff
@@ -0,0 +1,41 @@
+
+#include "system.h"
+#include "adc.h"
+#include "nuts_bolts.h"
+
+#define ADMUX_SELECTION_MASK    0x7
+#define MUX5_MASK               0x8
+
+uint16_t adc_read_channel(uint8_t channel) {
+  // Set ADC reference
+  ADMUX = (1 << REFS0);
+
+  // Enable ADC, Prescaler: 128x
+  ADCSRA = (1 << ADEN) | (1 << ADPS2) | (1 << ADPS1) | (1 << ADPS0);
+
+  // Select Channel
+  ADMUX |= (channel & ADMUX_SELECTION_MASK);
+
+  // Select Channel - ADCSRB, MUX5
+  if (channel & MUX5_MASK) {
+    bit_true(ADCSRB, MUX5_BIT_POS);
+  } else {
+    bit_false(ADCSRB, MUX5_BIT_POS);
+  }
+
+  // Enable ADC capture
+  ADCSRA |= (1 << ADSC);
+
+  // Enable capture of ADC
+  while(ADCSRA & (1 << ADSC));
+
+  // Stote the result
+  uint16_t ret = ADC;
+  
```

Comment:
Good point - the data sheet says the following:

> Special care should be taken when changing differential channels. Once a differential channel has been selected, the stage may take as much as 125μs to stabilize to the new value. Thus conversions should not be started within the first 125μs after selecting a new differential channel. Alternatively, conversion results obtained within this period should be discarded. The same settling time should be observed for the first differential conversion after changing ADC reference (by changing the REFS1:0 bits in ADMUX).

I will compensate for this by moving ADMUX = (1 << REFS0) to an adc_init() function in main.c. A believe that the time delay between adc_init and the first call to adc_read_channel will be at least 125 us.